### PR TITLE
Documentation for SSO using OpenID connect

### PR DESCRIPTION
**Pull request by apurvabhide**
@rahulbajaj0509 @gnurag can you please provide your inputs for this documentation?

### CODE REVIEW

**Comment by ekohl** [comment]
Please also update it in nightly so we don't lose it.

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (ekohl) [inline _includes/manuals/1.24/5.7.7_openid_sso.md]: I would indent these to make it clear they're part of the previous command.

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (ekohl) [inline _includes/manuals/1.24/5.7.7_openid_sso.md]: Does this modify the Apache configuration? If you rerun the installer after, does this revert the installation?

Comment (rabajaj0509) [reply _includes/manuals/1.24/5.7.7_openid_sso.md]: Yes, good catch! once installer is run again it will revert the changes back. I think a puppet module will be required for this, right @ekohl ? Should we add a note here for now or leave that to the handling of puppet module and merge this PR irrespective of that change?

What do you think @ekohl ?

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (rabajaj0509) [inline _includes/manuals/1.24/5.7.7_openid_sso.md]: nit: s / `https://keycloak.org` / `https://keycloak.example.com`

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (rabajaj0509) [inline _includes/manuals/1.24/5.7.7_openid_sso.md]: Can we use `OpenID provider` here, instead of IDP

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (rabajaj0509) [inline _includes/manuals/1.24/5.7.7_openid_sso.md]: same. `OpenID provider` instead of IDP

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (rabajaj0509) [inline _includes/manuals/1.24/5.7.7_openid_sso.md]: same. `OpenID provider` instead of IDP

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (rabajaj0509) [inline _includes/manuals/1.24/5.7.7_openid_sso.md]: same. `OpenID provider` instead of IDP

**Review by rabajaj0509 — CHANGES_REQUESTED** [review]
Just added a few in-line comments but overall this looks great! Thanks for all your efforts @apurvabhide.

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (ekohl) [reply _includes/manuals/1.24/5.7.7_openid_sso.md]: Actually, I think this is correct since it's referring to the project. There are other OIDC providers.

Comment (rabajaj0509) [reply _includes/manuals/1.24/5.7.7_openid_sso.md]: oh! my bad.

**Comment by apurvabhide** [comment]
> Please also update it in nightly so we don't lose it.

@ekohl , I am not quite aware of the nightly, could you please elaborate it?

**Comment by ekohl** [comment]
> @ekohl , I am not quite aware of the nightly, could you please elaborate it?

You now made all the changes to `/1.24` but we consider `/nightly` the next version. That means for 1.25 we will copy `/nightly` to `/1.25`. If your changes are only in 1.24, it means future releases won't contain it.

**Comment by apurvabhide** [comment]
> > @ekohl , I am not quite aware of the nightly, could you please elaborate it?
> 
> You now made all the changes to `/1.24` but we consider `/nightly` the next version. That means for 1.25 we will copy `/nightly` to `/1.25`. If your changes are only in 1.24, it means future releases won't contain it.

@ekohl , I have added it in the `/nightly` and removed it from `/1.24`.

**Comment by apurvabhide** [comment]
> > > @ekohl , I am not quite aware of the nightly, could you please elaborate it?
> > 
> > 
> > You now made all the changes to `/1.24` but we consider `/nightly` the next version. That means for 1.25 we will copy `/nightly` to `/1.25`. If your changes are only in 1.24, it means future releases won't contain it.
> 
> @ekohl , I have added it in the `/nightly` and removed it from `/1.24`.

@ekohl, I have added these changes in `/1.24` also. Apologies for the confusion.

**Review by ntkathole — COMMENTED** [review]
Overall PR looks good to me! Can we mention about default options to setup in hammer config for oauth/basic auth as well?

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (ntkathole) [inline _includes/manuals/1.24/5.7.7_openid_sso.md]: opensc is specifically required for smart card authentication. Can we remove this rpm for here and may be include in smartcard auth documentation?

Comment (rabajaj0509) [reply _includes/manuals/1.24/5.7.7_openid_sso.md]: perfect! thanks a lot for reviewing @ntkathole

Comment (apurvabhide) [reply _includes/manuals/1.24/5.7.7_openid_sso.md]: > Overall PR looks good to me! Can we mention about default options to setup in hammer config for oauth/basic auth as well?

@ntkathole, Thank you for reviewing, I would like to add hammer configuration options in another PR.

Comment (apurvabhide) [reply _includes/manuals/1.24/5.7.7_openid_sso.md]: > opensc is specifically required for smart card authentication. Can we remove this rpm for here and may be include in smartcard auth documentation?

@ntkathole, done, I have removed `opensc` from this command.

**Review by ntkathole — APPROVED** [review]
Thanks @apurvabhide @rahulbajaj0509 Documentation wise ACK. I am leaving syntax and style checking over reviewers.

**Review by rabajaj0509 — APPROVED** [review]
@apurvabhide thanks for all the efforts!

**Comment by ekohl** [comment]
This is incompatible with the installer. That doesn't have to block merging documentation but I do think it requires a warning.

**Comment by apurvabhide** [comment]
> This is incompatible with the installer. That doesn't have to block merging documentation but I do think it requires a warning.

@ekohl, done, I have added warning for the installer.

**Comment by ehelms** [comment]
If I recall, the original plan was for Keycloak users to make use of the Apache puppet modules purge configs being set to false to prevent the installer from clearing this configuration out:

https://github.com/puppetlabs/puppetlabs-apache/blob/31c14cf50251f7692a768b3e3d03d738613baa9b/manifests/init.pp#L294

**Comment by ekohl** [comment]
When we want to switch over from passenger to puma we actually rely on that setting to clear up passenger configs. While initially it won't harm them as long as the package is installed, the real cleanup (`yum remove`) will break things. I consider setting that to true an unsupported setup

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (ekohl) [inline _includes/manuals/1.24/5.7.7_openid_sso.md]: At least on CentOS 7.7 `python2-requests-oauthlib` can be dropped because it's pulled in automatically.

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (ekohl) [inline _includes/manuals/1.24/5.7.7_openid_sso.md:36]: This appears to fail for me. First I had to install python-lxml:
```
[root@68e8ec694948 /]# keycloak-httpd-client-install --help
Traceback (most recent call last):
  File "/usr/bin/keycloak-httpd-client-install", line 8, in <module>
    from lxml import etree
ImportError: No module named lxml
```

After that it doesn't recognize `-t`

```
[root@68e8ec694948 /]# export KEYCLOAK_URL=https://keycloak.example.com
[root@68e8ec694948 /]# export KEYCLOAK_REALM=ssl-realm
[root@68e8ec694948 /]# export KEYCLOAK_USER=admin
[root@68e8ec694948 /]# keycloak-httpd-client-install --app-name foreman-openidc --keycloak-server-url $KEYCLOAK_URL --keycloak-admin-username $KEYCLOAK_USER --keycloak-realm $KEYCLOAK_REALM --keycloak-admin-realm master --keycloak-auth-role root-admin -t openidc -l /users/extlogin -d
usage: keycloak-httpd-client-install [-h] [--no-root-check] [-v] [-d]
                                     [--show-traceback] [--log-file LOG_FILE]
                                     --app-name APP_NAME [--force]
                                     [--permit-insecure-transport]
                                     [--tls-verify TLS_VERIFY]
                                     [--template-dir TEMPLATE_DIR]
                                     [--httpd-dir HTTPD_DIR] -r KEYCLOAK_REALM
                                     -s KEYCLOAK_SERVER_URL
                                     [-a {root-admin,realm-admin,anonymous}]
                                     [-u KEYCLOAK_ADMIN_USERNAME]
                                     [-P KEYCLOAK_ADMIN_PASSWORD_FILE]
                                     [-p DEPRECATED_KEYCLOAK_ADMIN_PASSWORD]
                                     [--keycloak-admin-realm KEYCLOAK_ADMIN_REALM]
                                     [--initial-access-token INITIAL_ACCESS_TOKEN]
                                     [--client-originate-method {descriptor,registration}]
                                     [--mellon-key-file MELLON_KEY_FILE]
                                     [--mellon-cert-file MELLON_CERT_FILE]
                                     [--mellon-hostname MELLON_HOSTNAME]
                                     [--mellon-https-port MELLON_HTTPS_PORT]
                                     [--mellon-root MELLON_ROOT]
                                     [--mellon-endpoint MELLON_ENDPOINT]
                                     [--mellon-entity-id MELLON_ENTITY_ID]
                                     [--mellon-idp-attr-name MELLON_IDP_ATTR_NAME]
                                     [--mellon-organization-name MELLON_ORGANIZATION_NAME]
                                     [--mellon-organization-display-name MELLON_ORGANIZATION_DISPLAY_NAME]
                                     [--mellon-organization-url MELLON_ORGANIZATION_URL]
                                     [-l MELLON_PROTECTED_LOCATIONS]
keycloak-httpd-client-install: error: unrecognized arguments: -t openidc
```

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (ekohl) [inline _includes/manuals/1.24/5.7.7_openid_sso.md]: I'd avoid personal names in examples. Safer to have something as `myuser`

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (ekohl) [inline _includes/manuals/1.24/5.7.7_openid_sso.md]:
```suggestion
hammer auth login oauth --two-factor --oidc-token-endpoint https://keycloak.example.com/token --oidc-authorization-endpoint https://keycloak.example.com/auth --oidc-client-id foreman-client --oidc-redirect-uri urn:ietf:wg:oauth:2.0:oob
```

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (rabajaj0509) [reply _includes/manuals/1.24/5.7.7_openid_sso.md:36]: @ekohl  which version of `mod_auth_openidc` and `keycloak-httpd-client-install`
 are you using?

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (rabajaj0509) [reply _includes/manuals/1.24/5.7.7_openid_sso.md]: @ekohl good catch! I checked for Fedora, it comes as an installing dependency for the `keycloak-httpd-client-install` package.

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (ekohl) [reply _includes/manuals/1.24/5.7.7_openid_sso.md:36]: Looks like currently CentOS ships the following:
* keycloak-httpd-client-install.noarch 0:0.8-1.el7
* mod_auth_openidc.x86_64 0:1.8.8-5.el7

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (gnurag) [inline _includes/manuals/1.24/5.7.7_openid_sso.md:33]: On line 20, Keycloak URL is set to https://sso.example.com. Please update this (or the previous one?) URL to maintain consistency.

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (gnurag) [reply _includes/manuals/1.24/5.7.7_openid_sso.md:36]: keycloak-httpd-client-install.noarch version `0:0.8-1.el7` could be the reason for failure. I reckon OpenID Connect support was added to keycloak-httpd-client-install in version 1.x. 

@rahulbajaj0509 Did you test this setup with a custom built keycloak-httpd-client-install?
@upadhyeammit FYI

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (gnurag) [inline _includes/manuals/1.24/5.7.7_openid_sso.md]: can delete the changes in Apache files added by this feature -> can purge the changes in Apache files added by keycloak-httpd-client-install

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (gnurag) [inline _includes/manuals/1.24/5.7.7_openid_sso.md]: For example, For example -> For example,

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (gnurag) [inline _includes/manuals/1.24/5.7.7_openid_sso.md]: Do we need to link to Keycloak project's website again? We've already provided its link in the introduction at the top. What convention do other manuals follow?

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (gnurag) [inline _includes/manuals/1.24/5.7.7_openid_sso.md:61]: Perhaps add a section to demonstrate SSO workflow for Foreman API.

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (gnurag) [inline _includes/manuals/1.24/5.7.7_openid_sso.md]: ++ Foreman will automatically redirect to Keycloak Sign-in page. If credentials are correct, it redirects to Foreman dashboard.

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (gnurag) [inline _includes/manuals/1.24/5.7.7_openid_sso.md]: s/this/This/

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (gnurag) [inline _includes/manuals/1.24/5.7.7_openid_sso.md]: @rahulbajaj0509 Just a note for hammer implementation: `urn:ietf:wg:oauth:2.0:oob` is a standard redirect string and the user need not remember it. This redirect string is always used when `--two-factor` is passed, so can this be made implicit?

Comment (gnurag) [inline _includes/manuals/1.24/5.7.7_openid_sso.md]: Does `--oidc-client-id foreman-client` in this line correspond to `keycloak-httpd-client-install --app-name foreman-openidc` from line 35? If they're same, perhaps use the same names to avoid confusion. If not, please ignore.

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (gnurag) [inline _includes/manuals/1.24/5.7.7_openid_sso.md]: Please read comment on `--oidc-client-id foreman-client` above.

**Review by gnurag — COMMENTED** [review]
In general, please also pay attention to using appropriate capitalization.

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (apurvabhide) [reply _includes/manuals/1.24/5.7.7_openid_sso.md:61]: @gnurag , as discussed with @rahulbajaj0509 , I will add it in another PR in 5.1 API section .

**Comment by apurvabhide** [comment]
> In general, please also pay attention to using appropriate capitalization.

@gnurag, Done, I have updated both the files with your suggestions.

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (rabajaj0509) [reply _includes/manuals/1.24/5.7.7_openid_sso.md:36]: @gnurag that is correct, keycloak-httpd-client-install package with version >= 1.x. supports OpenID Connect.

@ekohl can we have this bit merged for now, so that we have some documentation about the feature rather than nothing? and then we can make improvements to it as when we face issues.
I am completely ok to not merge this if you believe so.

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (ekohl) [reply _includes/manuals/1.24/5.7.7_openid_sso.md]: Could you have a look at this?
```suggestion
# yum install mod_auth_openidc keycloak-httpd-client-install -y
```

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (ekohl) [reply _includes/manuals/1.24/5.7.7_openid_sso.md:36]: I would add a note that you need at least version 1.x and the one in base is too old.

Comment (rabajaj0509) [reply _includes/manuals/1.24/5.7.7_openid_sso.md:36]: Makes sense.

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (apurvabhide) [reply _includes/manuals/1.24/5.7.7_openid_sso.md]: @ekohl , I have updated this command.

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (apurvabhide) [reply _includes/manuals/1.24/5.7.7_openid_sso.md:36]: I have added a note for keycloak-httpd-client-install package version.

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (ekohl) [inline _includes/manuals/1.24/5.7.7_openid_sso.md]:
```suggestion
An OpenID Provider(OP) implements a Single Sign-on (SSO) using an OpenID Connect (OIDC) protocol. OIDC verifies the identity of the End-user depending on the given authentication details. In this example, [Keycloak](https://www.keycloak.org/) is used as the OpenID provider, and Foreman is the Relying Party (RP).
```

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (ekohl) [inline _includes/manuals/1.24/5.7.7_openid_sso.md]:
```suggestion
* A working installation of OpenID provider, for example, Keycloak at https://keycloak.example.com, which uses the OIDC protocol. Ensure that the Keycloak server is running.
```

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (ekohl) [inline _includes/manuals/1.24/5.7.7_openid_sso.md]:
```suggestion
Foreman will automatically redirect to Keycloak Sign-in page. If credentials are correct, it redirects to the Foreman dashboard.
```

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (ekohl) [inline _includes/manuals/1.24/5.7.7_openid_sso.md]: We can use console syntax highlighting (and lower is another block which can use this)
```suggestion
```console
```

**Review by ekohl — APPROVED** [review]
Small nitpicks that would be nice to fix but overall :+1: to merging.

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (rabajaj0509) [inline _includes/manuals/1.24/5.7.7_openid_sso.md]: I think the parameter is `RS256`

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (apurvabhide) [reply _includes/manuals/1.24/5.7.7_openid_sso.md]: For the indentation purpose, I have added one space before this sentence.

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (apurvabhide) [reply _includes/manuals/1.24/5.7.7_openid_sso.md]: done, I have updated this.

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (apurvabhide) [reply _includes/manuals/1.24/5.7.7_openid_sso.md]: done, I have added ```console for the console syntax highlighting.

**Comment by rabajaj0509** [comment]
Thanks @apurvabhide for all the hard work! merging :)

**Comment by apurvabhide** [comment]
> Thanks @apurvabhide for all the hard work! merging :)

My pleasure :) 
Thank you all, for reviewing it.

File: _includes/manuals/1.24/5.7.7_openid_sso.md
Comment (ekohl) [inline _includes/manuals/1.24/5.7.7_openid_sso.md:41]: I had a look at this and this image is not included leading to a HTTP 404. This is true for other images as well.

Comment (apurvabhide) [reply _includes/manuals/1.24/5.7.7_openid_sso.md:41]: @ekohl , apologies for the late reply. Could you please help me with the image folder, should I add images in the /static/images/ directory?
Thanks

**Pull request merged**
